altair/utils/schemapi.py

```diff
@@ -348,12 +445,27 @@ def _todict(obj, context):
         return obj
 
 
-def _resolve_references(schema, root=None):
-    """Resolve schema references."""
-    resolver = jsonschema.RefResolver.from_schema(root or schema)
-    while "$ref" in schema:
-        with resolver.resolving(schema["$ref"]) as resolved:
-            schema = resolved
+def _resolve_references(
+    schema: Dict[str, Any], rootschema: Optional[Dict[str, Any]] = None
+) -> Dict[str, Any]:
+    """Resolve schema references until there is no $ref anymore
+    in the top-level of the dictionary.
+    """
+    if _use_referencing_library():
+        registry = _get_referencing_registry(rootschema or schema)
```

Comment:
I'll take a closer look at the whole PR tomorrow, but just because this is the first thing that pops out -- do you know why you do all this in the first place (pre-resolving references)? Doing so shouldn't be needed, and in fact is discouraged in the specification (basically because it's impossible in the general case) -- but it shouldn't be needed unless your schemas don't really have valid `$ref`s (in which case ideally you'd fix that in the schema, though maybe you don't control the Vega schemas?)

Reply:
I don't know why this happens here. I do see your point that this looks redundant at forst glance. So far I just tried to keep the functionality the same of this function. I won't get to this before mid of next week but then I can investigate what happens to our test suite and the code generation scripts if I remove this.

Reply:
I looked into this and we need the recursive resolving not for the schema validation but for other things such as automatic code generation, dynamic population of methods on classes, etc. For example, class `alt.Y` needs to follow the json schema which can be found at `{'$ref': '#/definitions/PositionFieldDef'}` in the full Vega-Lite schema. We therefore only store that dictionary as reference on the Python class. However, in some parts of the code (e.g. https://github.com/altair-viz/altair/blob/main/altair/utils/schemapi.py#L1136), we need to know what is behind that references as we want to iterate through the schema in Python.

As part of a bigger rework, we could try to simplify this logic but that is outside of the scope of this PR.